fiona/_transform.pyx

```diff
@@ -127,6 +128,9 @@ def _transform_geom(
     cdef void *dst_ogr_geom = NULL
     cdef int i
 
+    if not isinstance(geom, Geometry):
+        geom = Geometry.from_dict(**geom)
+
```

Comment:
This is perhaps better done in transform.py. We could strictly require Geometry in the extension module since it isn't in the public API.